### DISCUSSION TITLE
How to accept Annotated types as well as normal types?

### DISCUSSION THREAD

**Discussion by superlopuh** [microsoft/pyright#7849] [Q&A] [answered]
Here's a minimal snippet:

``` python
from typing import Annotated, TypeAlias


class MyClass: ...


def my_function(arg: type[MyClass]) -> None: ...


MyTypealias: TypeAlias = Annotated[MyClass, 1]

# works
my_function(MyClass)
# doesn't work
my_function(MyTypealias)
```

The error:
```
.../hello.py:15:13 - error: Argument of type "MyTypealias" cannot be assigned to parameter "arg" of type "type[MyClass]" in function "my_function"
    Type "type[Annotated]" is incompatible with type "type[MyClass]" (reportArgumentType)
```

I'm surprised, I thought that Annotated was supposed to be transparent wrt the type system. The issue is that `type[Annotated]` seems to not be something I can spell, if I add it to the list of types accepted by `my_function`, I get two errors, the existing one, as well as `Expected type arguments for generic class "Annotated"`.

Is there any way to make this work?

**Comment by erictraut** [comment, marked as answer]
`Annotated` is intended to be used in type expressions (i.e. type annotations). You're using it here in a value expression. See [this issue](https://github.com/microsoft/pyright/issues/7238) for more details.

What are you doing in `my_function` with the value you pass to it? The value `MyTypealias` is not a valid class object, so `my_function` is not going to be able to use it as a normal `type`. For example, if there is a class variable defined on `MyClass`, you won't be able to access it using `arg.my_class_var`.

**Reply by superlopuh** [reply]
Ah, I see, thank you.

We use this as part of a declarative constraint system. `my_function` is a dataclass field descriptor that leverages the type system for static/IDE checking, and we use Annotated to add additional constraints that the type system doesn't support, but we can in our custom `verify` function. This pattern is currently used to enforce a same-type constraint when multiple fields must have the same type. I'll try to find a different API for this.

**Reply by erictraut** [reply]
You may be interested in the draft PEP discussed in [this thread](https://discuss.python.org/t/typeform-spelling-for-a-type-annotation-object-at-runtime/51435/8).
